**IMPORT DATA**

In [2]:
import pandas as pd
data=pd.read_csv(r"C:\Users\Personal Camputer\OneDrive\Desktop\data\HRDataset_v14.csv")
pd.set_option("display.max_columns",None)
data

,Employee_Name,EmpID,MarriedID,MaritalStatusID,GenderID,EmpStatusID,DeptID,PerfScoreID,FromDiversityJobFairID,Salary,Termd,PositionID,Position,State,Zip,DOB,Sex,MaritalDesc,CitizenDesc,HispanicLatino,RaceDesc,DateofHire,DateofTermination,TermReason,EmploymentStatus,Department,ManagerName,ManagerID,RecruitmentSource,PerformanceScore,EngagementSurvey,EmpSatisfaction,SpecialProjectsCount,LastPerformanceReview_Date,DaysLateLast30,Absences
0,"Adinolfi, Wilson K",10026,0,0,1,1,5,4,0,62506,0,19,Production Technician I,MA,1960,07/10/83,M,Single,US Citizen,No,White,7/5/2011,NaN,N/A-StillEmployed,Active,Production,Michael Albert,22.0,LinkedIn,Exceeds,4.60,5,0,1/17/2019,0,1
1,"Ait Sidi, Karthikeyan",10084,1,1,1,5,3,3,0,104437,1,27,Sr. DBA,MA,2148,05/05/75,M,Married,US Citizen,No,White,3/30/2015,6/16/2016,career change,Voluntarily Terminated,IT/IS,Simon Roup,4.0,Indeed,Fully Meets,4.96,3,6,2/24/2016,0,17
2,"Akinkuolie, Sarah",10196,1,1,0,5,5,3,0,64955,1,20,Production Technician II,MA,1810,09/19/88,F,Married,US Citizen,No,White,7/5/2011,9/24/2012,hours,Voluntarily Terminated,Production,Kissy Sullivan,20.0,LinkedIn,Fully Meets,3.02,3,0,5/15/2012,0,3
3,"Alagbe,Trina",10088,1,1,0,1,5,3,0,64991,0,19,Production Technician I,MA,1886,09/27/88,F,Married,US Citizen,No,White,1/7/2008,NaN,N/A-StillEmployed,Active,Production,Elijiah Gray,16.0,Indeed,Fully Meets,4.84,5,0,1/3/2019,0,15
4,"Anderson, Carol",10069,0,2,0,5,5,3,0,50825,1,19,Production Technician I,MA,2169,09/08/89,F,Divorced,US Citizen,No,White,7/11/2011,9/6/2016,return to school,Voluntarily Terminated,Production,Webster Butler,39.0,Google Search,Fully Meets,5.00,4,0,2/1/2016,0,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
306,"Woodson, Jason",10135,0,0,1,1,5,3,0,65893,0,20,Production Technician II,MA,1810,05/11/85,M,Single,US Citizen,No,White,7/7/2014,NaN,N/A-StillEmployed,Active,Production,Kissy Sullivan,20.0,LinkedIn,Fully Meets,4.07,4,0,2/28/2019,0,13
307,"Ybarra, Catherine",10301,0,0,0,5,5,1,0,48513,1,19,Production Technician I,MA,2458,05/04/82,F,Single,US Citizen,No,Asian,9/2/2008,9/29/2015,Another position,Voluntarily Terminated,Production,Brannon Miller,12.0,Google Search,PIP,3.20,2,0,9/2/2015,5,4
308,"Zamora, Jennifer",10010,0,0,0,1,3,4,0,220450,0,6,CIO,MA,2067,08/30/79,F,Single,US Citizen,No,White,4/10/2010,NaN,N/A-StillEmployed,Active,IT/IS,Janet King,2.0,Employee Referral,Exceeds,4.60,5,6,2/21/2019,0,16
309,"Zhou, Julia",10043,0,0,0,1,3,3,0,89292,0,9,Data Analyst,MA,2148,02/24/79,F,Single,US Citizen,No,White,3/30/2015,NaN,N/A-StillEmployed,Active,IT/IS,Simon Roup,4.0,Employee Referral,Fully Meets,5.00,3,5,2/1/2019,0,11


**Question one**

In [5]:
data["DOB"]=pd.to_datetime(data["DOB"])
today = pd.Timestamp.today()
data["age"]=((today- data["DOB"]).dt.days/365).astype(int)
data[["DOB","age"]]

C:\Users\Personal Camputer\AppData\Local\Temp\ipykernel_52664\4165605638.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data["DOB"]=pd.to_datetime(data["DOB"])


,DOB,age
0,1983-07-10,43
1,2075-05-05,-48
2,1988-09-19,37
3,1988-09-27,37
4,1989-09-08,36
...,...,...
306,1985-05-11,41
307,1982-05-04,44
308,1979-08-30,46
309,1979-02-24,47


In [6]:
data["group by age"]=pd.cut(
    data["age"],
    bins=[0, 25, 40,60,150],
    labels=["<25","25-40","40-60","60<"],
    right=False
)
data[["age","group by age"]],


(     age group by age
 0     43        40-60
 1    -48          NaN
 2     37        25-40
 3     37        25-40
 4     36        25-40
 ..   ...          ...
 306   41        40-60
 307   44        40-60
 308   46        40-60
 309   47        40-60
 310   47        40-60
 
 [311 rows x 2 columns],)

In [ ]:
data["group by age"].value_counts()

group by age
40-60    138
25-40     73
<25        0
60<        0
Name: count, dtype: int64


In [ ]:
avg_absence=(
    data.groupby("Position")["Absences"].mean().astype(int)
    .sort_values(ascending=False)
)
print(avg_absence)
print("______________________________________")


Position
Shared Services Manager         20
Director of Sales               19
Data Architect                  17
CIO                             16
IT Director                     15
Director of Operations          15
IT Manager - Infra              13
Data Analyst                    12
Principal Data Architect        12
Senior BI Developer             12
IT Support                      12
Sr. DBA                         12
Area Sales Manager              11
BI Developer                    11
Production Technician II        10
IT Manager - DB                 10
Sr. Network Engineer            10
President & CEO                 10
Production Technician I         10
Software Engineer                9
Accountant I                     8
Administrative Assistant         8
Production Manager               8
BI Director                      8
Sales Manager                    8
Network Engineer                 6
Database Administrator           6
IT Manager - Support             4
Software En

In [7]:
avg_salary=(
    data.groupby("Position")["Salary"].mean()
    .sort_values(ascending=False)
)
print(avg_salary)

Position
President & CEO                 250000.000000
CIO                             220450.000000
Director of Sales               180000.000000
IT Director                     178000.000000
Director of Operations          170500.000000
IT Manager - Infra              157000.000000
Data Architect                  150290.000000
IT Manager - DB                 144959.500000
IT Manager - Support            138888.000000
Principal Data Architect        120000.000000
BI Director                     110929.000000
Database Administrator          108499.600000
Enterprise Architect            103613.000000
Sr. Accountant                  102859.000000
Sr. DBA                         102234.000000
Software Engineer                96719.200000
BI Developer                     95465.000000
Sr. Network Engineer             93070.800000
Shared Services Manager          93046.000000
Data Analyst                     89932.571429
Data Analyst                     88527.000000
Senior BI Developer      

**Question three**

In [65]:
terminate=data["DateofTermination"].notna()
percentage=terminate.mean()*100
print(f"percentage of employees who left: {percentage:.2f}%")


percentage of employees who left: 33.44%


In [66]:
leftcompany = data[data["DateofTermination"].notna()]
single = (leftcompany["MaritalDesc"]=="Single").sum()
married = (leftcompany["MaritalDesc"] == "Married").sum()

print(f"Singles who left the company: {single}")
print(f"Marrieds who turnedover: {married}")

Singles who left the company: 36
Marrieds who turnedover: 47


Based on the resuls, yes, single employees left the company more often than married employees. This suggests that employee turnover was higher among single employees.we should notice that, this result is based on the number of employees who left and doesnt necessarily mean that being single is the cause of higher turnover.

**Question four**

In [74]:
import pandas as pd

data["DateofHire"] = pd.to_datetime(data["DateofHire"])
data["DateofTermination"] = pd.to_datetime(data["DateofTermination"])

leftdf = data[data["DateofTermination"].notna()]
result = leftdf["DateofTermination"] - leftdf["DateofHire"]
#per year
average_years = result.dt.days.mean() / 365
print(f"Average years of employees who work before they left: {average_years:.2f} years")

#per day
#average_days = result.dt.days.mean()
#print(f"Average day of employees who work befor they left: {average_days:.2f} days")

Average years of employees who work before they left: 3.46 years


The average tenure of employees who left the company was 3.46 years. This means that employees stayed with the company for about 4 years on average before leaving.

In [ ]:
new=data["DateofHire"].dt.year.value_counts()
print(f"highst hiring rate in each year:{new}")

highst hiring rate in range:DateofHire
2011    83
2014    60
2012    45
2013    44
2015    36
2016    14
2010     9
2009     7
2017     6
2008     3
2007     2
2018     1
2006     1
Name: count, dtype: int64


**Question five**

In [16]:
data["LeftCompany"] = data["DateofTermination"].notna()

result = data.groupby("LeftCompany")["EmpSatisfaction"].mean()

print(result)

#EmpsWhoLeft = data[data["DateofTermination"].notna()]
#left_sat = EmpsWhoLeft["EmpSatisfaction"].mean()

#EmpsWhoStay = data[data["DateofTermination"].isna()]
#stay_sat = EmpsWhoStay["EmpSatisfaction"].mean()

#print(f"Average job satisfaction of employees who left: {left_sat:.2f}")
#print(f"Average job satisfaction of employees who stayed: {stay_sat:.2f}")

LeftCompany
False    3.893720
True     3.884615
Name: EmpSatisfaction, dtype: float64


The average job satisfaction of employees who left and those who stayed was almost the same. There was no noticeable difference between the two groups.

In [4]:

left = data[data["DateofTermination"].notna()]

left["TermReason"].value_counts().head(5)

TermReason
Another position    20
unhappy             14
more money          11
career change        9
hours                8
Name: count, dtype: int64